[mlir] Require folders to produce Values of same type

matthias-springer opened the pull request:
This commit adds extra assertions to `OperationFolder` and `OpBuilder` to ensure that the types of the folded SSA values match with the result types of the op. There used to be checks that discard the folded results if the types do not match. This commit makes these checks stricter and turns them into assertions.

Discarding folded results with the wrong type (without failing explicitly) can hide bugs in op folders. Two such bugs became apparent in MLIR (and some more in downstream projects) and are fixed with this change.

Note: The existing type checks were introduced in https://reviews.llvm.org/D95991.

Migration guide: If you see failing assertions (`folder produced value of incorrect type`; make sure to run with assertions enabled!), run with `-debug` or dump the operation right before the failing assertion. This will point you to the op that has the broken folder. A common mistake is a mismatch between static/dynamic dimensions (e.g., input has a static dimension but folded result has a dynamic dimension).

kuhar reviewed (approved):
LGTM

jeanPerier reviewed (approved):
Thanks for the fix in flang, LGTM

matthias-springer committed:
[mlir] Require folders to produce Values of same type

This commit adds extra assertions to `OperationFolder` and `OpBuilder` to ensure that the types of the folded SSA values match with the result types of the op. There used to be checks that discard the folded results if the types do not match. This commit makes these checks stricter and turns them into assertions.

Discarding folded results with the wrong type can hide bugs in op folders. Two such bugs became apparent and are fixed with this change.

Note: The existing type checks were introduced in https://reviews.llvm.org/D95991.

BEGIN_PUBLIC
No public commit message needed for presubmit.
END_PUBLIC
